### PR TITLE
chore: add simproc to simplify `a % b` using omega

### PR DESCRIPTION
**Pull request by bollu**
We try to apply `mod_eq_of_lt`.

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
@bollu, can you review this?

File: SSA/Experimental/SimprocBv.lean [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  (x + y).toNat = (x.toNat + y.toNat) % 2 ^ w := by
+  simp
+
+/-- info: 'eg₂' depends on axioms: [propext, Quot.sound] -/
+#guard_msgs in #print axioms eg₂
+
+theorem eg₄ (x y z : BitVec w)
+  (h₂ : y.toNat + z.toNat < 2^w)
+  (h : x.toNat * (y.toNat + z.toNat) < 2^w) :
+  (x * (y + z)).toNat = x.toNat * (y.toNat + z.toNat) := by
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
theorem eg₄ (x y z : BitVec w)
  (h₂ : y.toNat + z.toNat < 2 ^ w)
  (h : x.toNat * (y.toNat + z.toNat) < 2 ^ w) :
  (x * (y + z)).toNat = x.toNat * (y.toNat + z.toNat) := by
```

File: SSA/Experimental/SimprocBv.lean [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+simproc↑ reduce_mod_eq_of_lt (_ % _) := fun e => reduceModEqOfLt e
+
+theorem eg₁ (x : BitVec w) : x.toNat % 2^w = x.toNat + 0:= by
+  simp
+
+/-- info: 'eg₁' depends on axioms: [propext] -/
+#guard_msgs in #print axioms eg₁
+
+theorem eg₂ (x y : BitVec w)  (h : x.toNat + y.toNat < 2^w) :
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
theorem eg₁ (x : BitVec w) : x.toNat % 2 ^ w = x.toNat + 0:= by
  simp

/-- info: 'eg₁' depends on axioms: [propext] -/
#guard_msgs in #print axioms eg₁

theorem eg₂ (x y : BitVec w)  (h : x.toNat + y.toNat < 2 ^ w) :
```

File: SSA/Experimental/SimprocBv.lean [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+       g.withContext (do
+          let hyps := (← getLocalHyps).toList
+          Lean.Elab.Tactic.Omega.omega hyps g {})
+       let eqProof ← mkAppM ``Nat.mod_eq_of_lt #[ltProof]
+       return .done { expr := x, proof? := eqProof : Result }
+     catch _ =>
+       return .visit { expr := e }
+  | _ => do
+     return .visit { expr := e  : Result }
+
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
     return .visit { expr := e : Result }

```

File: SSA/Experimental/SimprocBv.lean [hunk 4]
```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2023 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddharth Bhat, Tobias Groser
+-/
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
Authors: Siddharth Bhat, Tobias Grosser
-/
```

File: SSA/Experimental/SimprocBv.lean [hunk 5]
```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2023 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
Copyright (c) 2023 Tobias Grosser, Siddharth Bhat, LLC. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
```

File: SSA/Experimental/SimprocBv.lean [hunk 6]
```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2023 Tobias Grosser, Siddharth Bhat, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
Copyright (c) 2023 Tobias Grosser, Siddharth Bhat. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
```

File: SSA/Experimental/SimprocBv.lean [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+Authors: Siddharth Bhat, Tobias Grosser
+-/
+import Lean
+
+open Lean Meta Elab Simp
+
+theorem Nat.mod_eq_sub {x y : Nat} (h : x ≥ y) (h' : x - y < y) :
+    x % y = x - y := by
+  rw [Nat.mod_eq_sub_mod h]
+  rw [Nat.mod_eq_of_lt h']
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
  rw [Nat.mod_eq_sub_mod h, Nat.mod_eq_of_lt h']
```

File: SSA/Experimental/SimprocBv.lean [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    if let .done res ← reduceModSub x n then
+      return .done res
+    return .continue
+  | _ => do
+     return .continue
+
+simproc↑ reduce_mod_omega (_ % _) := fun e => reduceMod e
+
+theorem eg₁ (x : BitVec w) : x.toNat % 2 ^ w = x.toNat + 0:= by
+  simp
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
theorem eg₁ (x : BitVec w) : x.toNat % 2 ^ w = x.toNat + 0 := by
  simp
```

File: SSA/Experimental/SimprocBv.lean [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  let ltTy := mkLTNat subTy n
+  let Step.done { expr := _, proof? := some ltProof} ← proofOmega ltTy
+    | return .continue
+  let eqProof ← mkAppM ``Nat.mod_eq_sub #[geProof, ltProof]
+  return .done { expr := subTy, proof? := eqProof : Result }
+
+@[inline] def reduceMod (e : Expr) : SimpM Step := do
+  match_expr e with
+  | HMod.hMod xTy nTy outTy  _inst x n =>
+    let natTy := mkConst ``Nat
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
  | HMod.hMod xTy nTy outTy _inst x n =>
    let natTy := mkConst ``Nat
```

File: SSA/Experimental/SimprocBv.lean [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  let instHSub := mkAppN (mkConst ``instHSub [lz]) #[nat, instSub]
+  mkAppN (mkConst ``HSub.hSub [lz, lz, lz]) #[nat, nat, nat, instHSub, x, y]
+
+@[inline] def proofOmega (ty : Expr) : SimpM Step := do
+  let proof : Expr ← mkFreshExprMVar ty
+  let proofMVar := proof.mvarId!
+  let some g ← proofMVar.falseOrByContra
+    | return .continue
+  try
+    g.withContext (do Lean.Elab.Tactic.Omega.omega (← getLocalHyps).toList g {})
```
Comment (bollu) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
  let g := proof.mvarId!
  let some g ← g.falseOrByContra
    | return .continue
  try
    g.withContext (do Lean.Elab.Tactic.Omega.omega (← getLocalHyps).toList g {})
```

Let's be consistent and name the goal `g` as lean often does.

File: SSA/Experimental/SimprocBv.lean [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+private def mkGENat (x y : Expr) : Expr := mkLENat y x
+
+private def mkSubNat (x y : Expr) : Expr :=
+  let lz := levelZero
+  let nat := mkConst ``Nat
+  let instSub := mkConst ``instSubNat
+  let instHSub := mkAppN (mkConst ``instHSub [lz]) #[nat, instSub]
+  mkAppN (mkConst ``HSub.hSub [lz, lz, lz]) #[nat, nat, nat, instHSub, x, y]
+
+@[inline] def proofOmega (ty : Expr) : SimpM Step := do
```
Comment (bollu) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
/-- Try to build a proof for `ty` by reduction to `omega`.
@[inline] def proveByOmega (ty : Expr) : SimpM Step := do
```

File: SSA/Experimental/SimprocBv.lean [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+  try
+    g.withContext (do Lean.Elab.Tactic.Omega.omega (← getLocalHyps).toList g {})
+  catch _ =>
+    return .continue
+  return .done { expr := ty, proof? := proof }
+
+-- x % n = x if x < n
+@[inline] def reduceModOfLt (x : Expr) (n : Expr) : SimpM Step := do
+  let ltTy := mkLTNat x n
+  let Step.done { expr := _, proof? := some p} ← proofOmega ltTy
```
Comment (bollu) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
  let Step.done { expr := _, proof? := some p} ← proveByOmega ltTy
```

File: SSA/Experimental/SimprocBv.lean [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    | return .continue
+  let eqProof ← mkAppM ``Nat.mod_eq_of_lt #[p]
+  return .done { expr := x, proof? := eqProof : Result }
+
+-- x % n = x - n if x >= n and x - n < n
+@[inline] def reduceModSub (x : Expr) (n : Expr) : SimpM Step := do
+  let geTy := mkGENat x n
+  let Step.done { expr := _, proof? := some geProof} ← proofOmega geTy
```
Comment (bollu) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
  let Step.done { expr := _, proof? := some geProof} ← proveByOmega geTy
```

File: SSA/Experimental/SimprocBv.lean [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    | return .continue
+  let subTy := mkSubNat x n
+  let ltTy := mkLTNat subTy n
+  let Step.done { expr := _, proof? := some ltProof} ← proofOmega ltTy
```
Comment (bollu) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
  let Step.done { expr := _, proof? := some ltProof} ← proveByOmega ltTy
```

**Review by bollu — COMMENTED** [review]
LGTM, some naming nits.

File: SSA/Experimental/SimprocBv.lean [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+private def mkSubNat (x y : Expr) : Expr :=
+  let lz := levelZero
+  let nat := mkConst ``Nat
+  let instSub := mkConst ``instSubNat
+  let instHSub := mkAppN (mkConst ``instHSub [lz]) #[nat, instSub]
+  mkAppN (mkConst ``HSub.hSub [lz, lz, lz]) #[nat, nat, nat, instHSub, x, y]
+
+/-- Try to build a proof for `ty` by reduction to `omega`.
+@[inline] def proveByOmega (ty : Expr) : SimpM Step := do
```
Comment (tobiasgrosser) [inline SSA/Experimental/SimprocBv.lean]:
```suggestion
/-- Try to build a proof for `ty` by reduction to `omega`. -/
@[inline] def proveByOmega (ty : Expr) : SimpM Step := do
```

**Pull request merged**